File: greathost.js
const EMAIL = process.env.GREATHOST_EMAIL || '';
const PASSWORD = process.env.GREATHOST_PASSWORD || '';
const CHAT_ID = process.env.CHAT_ID || '';
const BOT_TOKEN = process.env.BOT_TOKEN || '';
const PROXY_URL = (process.env.PROXY_URL || "").trim();

// 核心改动：换成 firefox 引擎
const { firefox } = require("playwright");
const https = require('https');

async function sendTelegramMessage(message) {
    return new Promise((resolve) => {
        const url = `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`;
        const data = JSON.stringify({ chat_id: CHAT_ID, text: message, parse_mode: 'HTML' });
        const options = { method: 'POST', headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(data) } };
        const req = https.request(url, options, (res) => {
            res.on('data', () => {});
            res.on('end', () => resolve());
        });
        req.on('error', () => resolve());
        req.write(data);
        req.end();
    });
}

(async () => {
    // === 严格保留你原始定义的 URL 变量 ===
    const GREATHOST_URL = "https://greathost.es";    
    const LOGIN_URL = `${GREATHOST_URL}/login`;
    const HOME_URL = `${GREATHOST_URL}/dashboard`;
    const BILLING_URL = `${GREATHOST_URL}/billing/free-servers`;
    
    let proxyStatusTag = "🌐 直连模式";
    let serverStarted = false;

    // --- 代理逻辑解析 ---
    let proxyData = null;
    if (PROXY_URL) {
        try {
            const cleanUrl = PROXY_URL.startsWith('socks') ? PROXY_URL : `socks5://${PROXY_URL}`;
            proxyData = new URL(cleanUrl);
            proxyStatusTag = `🔒 代理模式 (${proxyData.host})`;
        } catch (e) {
            console.error("❌ PROXY_URL 解析失败:", e.message);
        }
    }

    let browser;
    try {
        console.log(`🚀 任务启动 | 引擎: Firefox | ${proxyStatusTag}`);
        
        // 1. 启动浏览器（不带参数）
        browser = await firefox.launch({ headless: true });

        // 2. 在创建上下文时，【一次性】注入代理服务器和认证信息
        // 这是 Playwright Node.js 官方文档定义的标准 SOCKS5 认证方式
        const contextOptions = {
            userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0',
            viewport: { width: 1280, height: 720 },
            locale: 'es-ES'
        };

        if (proxyData) {
            contextOptions.proxy = {
                server: `socks5://${proxyData.host}`,
                username: proxyData.username || '',
                password: proxyData.password || ''
            };
        }

        const context = await browser.newContext(contextOptions);

        // 3. 创建页面
        const page = await context.newPage();

        // --- 完整保留你原来的指纹抹除 ---
        await page.addInitScript(() => {
            Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
        });

        // --- 1. 代理检测（熔断逻辑） ---
        if (proxyData) {
            console.log("🌍 [Check] 正在检测代理出口 IP...");
            try {
                await page.goto("https://api.ipify.org?format=json", { timeout: 30000 });
                const ipInfo = JSON.parse(await page.innerText('body'));
                console.log(`✅ 当前出口 IP: ${ipInfo.ip}`);
            } catch (e) {
                await sendTelegramMessage(`🚨 <b>GreatHost 代理异常</b>\n原因: ${e.message}`);
                throw new Error("Proxy Check Failed"); 
            }
        }

        // --- 2. 登录流程（严格按照你源代码的步骤） ---
        console.log("🔑 正在登录...");
        await page.goto(LOGIN_URL, { waitUntil: "domcontentloaded" });
        await page.fill('input[name="email"]', EMAIL);
        await page.fill('input[name="password"]', PASSWORD);
        await Promise.all([
            page.click('button[type="submit"]'),
            page.waitForNavigation({ waitUntil: "networkidle" }),
        ]);
        
        if (page.url().includes('login')) {
            throw new Error("登录失败，请检查账号密码");
        }
        console.log("✅ 登录成功！");

        // --- 3. 自动开机逻辑（完整保留） ---
        console.log("📊 检查服务器状态...");
        await page.goto(HOME_URL, { waitUntil: "networkidle" });
        const offlineIndicator = page.locator('span.badge-danger, .status-offline').first();
        if (await offlineIndicator.isVisible()) {
            console.log("⚠️ 检测到服务器离线，尝试启动...");
            const startBtn = page.locator('button:has-text("Start"), .btn-start').first();
            if (await startBtn.isVisible()) {
                await startBtn.click();
                serverStarted = true;
                await page.waitForTimeout(3000);
            }
        }

        // --- 4. 续期流程（还原原始点击路径和报表逻辑） ---
        console.log("🔍 进入 Billing 页面...");
        await page.goto(BILLING_URL, { waitUntil: "networkidle" });

        // 原版点击 View Details
        await page.getByRole('link', { name: 'View Details' }).first().click();
        await page.waitForNavigation({ waitUntil: "networkidle" });
        
        const serverId = page.url().split('/').pop() || 'unknown';
        const timeSelector = '#accumulated-time';

        // 捕获续期前时长
        const beforeHoursText = await page.textContent(timeSelector);
        const beforeHours = parseInt(beforeHoursText.replace(/[^0-9]/g, '')) || 0;

        const renewBtn = page.locator('#renew-free-server-btn');
        const btnContent = await renewBtn.innerHTML();

        // 完整保留你原来的报告生成函数
        const getReport = (icon, title, hours, detail) => {
            return `${icon} <b>GreatHost ${title}</b>\n\n` +
                   `🆔 <b>服务器ID:</b> <code>${serverId}</code>\n` +
                   `⏰ <b>当前时长:</b> ${hours}h\n` +
                   `🚀 <b>开机状态:</b> ${serverStarted ? '✅ 已触发开机' : '运行中'}\n` +
                   `🌐 <b>连接模式:</b> ${proxyStatusTag}\n` + 
                   `💡 <b>详情:</b> ${detail}`;
        };

        if (btnContent.includes('Wait')) {
            const waitMatch = btnContent.match(/\d+/);
            const waitTime = waitMatch ? waitMatch[0] : "??";
            await sendTelegramMessage(getReport('⏳', '续期冷却中', beforeHours, `还需等待 ${waitTime} 分钟`));
            return;
        }

        // --- 5. 执行续期（原始模拟逻辑） ---
        console.log("⚡ 执行续期操作...");
        await page.mouse.wheel(0, 200); 
        await page.waitForTimeout(2000);
        await renewBtn.click({ force: true });

        // --- 6. 最终校验（原始同步逻辑） ---
        console.log("⏳ 等待 20 秒处理数据...");
        await page.waitForTimeout(20000);
        await page.reload();
        
        const afterHoursText = await page.textContent(timeSelector);
        const afterHours = parseInt(afterHoursText.replace(/[^0-9]/g, '')) || 0;

        if (afterHours > beforeHours) {
            await sendTelegramMessage(getReport('🎉', '续期成功', afterHours, `时长从 ${beforeHours}h 增加`));
        } else {
            await sendTelegramMessage(getReport('✅', '已检查', afterHours, '目前时长充足，无需重复续期'));
        }

    } catch (err) {
        console.error("❌ 错误详情:", err);
        if (!err.message.includes("Proxy Check Failed")) {
            await sendTelegramMessage(`🚨 <b>GreatHost 脚本崩溃</b>\n错误: <code>${err.message}</code>`);
        }
    } finally {
        if (browser) {
            console.log("🧹 关闭浏览器...");
            await browser.close();
        }
    }
})();

File: greathost.py
import os, re, time, random, requests, json
from datetime import datetime, timezone
from urllib.parse import urlparse
from zoneinfo import ZoneInfo

from seleniumwire import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

# ================= 配置区 =================
EMAIL = os.getenv("GREATHOST_EMAIL", "")
PASSWORD = os.getenv("GREATHOST_PASSWORD", "")
TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN", "")
TELEGRAM_CHAT_ID = os.getenv("TELEGRAM_CHAT_ID", "")
# 代理sock5可留空
PROXY_URL = os.getenv("PROXY_URL", "")

# 目标服务器名
TARGET_NAME_CONFIG = os.getenv("TARGET_NAME", "loveMC")

STATUS_MAP = {
    "running": ["🟢", "Running"],
    "starting": ["🟡", "Starting"],
    "stopped": ["🔴", "Stopped"],
    "offline": ["⚪", "Offline"],
    "suspended": ["🚫", "Suspended"]
}

# ================= 工具函数 =================
def now_shanghai():
    return datetime.now(ZoneInfo("Asia/Shanghai")).strftime('%Y/%m/%d %H:%M:%S')

def calculate_hours(date_str):
    try:
        if not date_str: return 0
        # 兼容处理带毫秒的格式
        clean_date = re.sub(r'\.\d+Z$', 'Z', date_str)
        expiry = datetime.fromisoformat(clean_date.replace('Z', '+00:00'))
        now = datetime.now(timezone.utc)
        diff = (expiry - now).total_seconds() / 3600
        return max(0, int(diff))
    except Exception as e:
        print(f"⚠️ 时间解析失败: {e}")
        return 0

def fetch_api(driver, url, method="GET"):
    script = f"return fetch('{url}', {{method:'{method}'}}).then(r=>r.json()).catch(e=>({{success:false,message:e.toString()}}))"
    res = driver.execute_script(script)
    print(f"📡 API 调用 [{method}] {url}")
    return res

def send_notice(kind, fields):
    """保持您要求的 TG 通知风格"""
    titles = {
        "renew_success": "🎉 <b>GreatHost 续期成功</b>",
        "maxed_out": "🈵 <b>GreatHost 已达上限</b>",
        "cooldown": "⏳ <b>GreatHost 还在冷却中</b>",
        "renew_failed": "⚠️ <b>GreatHost 续期未生效</b>",
        "error": "🚨 <b>GreatHost 脚本报错</b>"
    }
    title = titles.get(kind, "‼️ <b>GreatHost 通知</b>")
    body = "\n".join([f"{e} {l}: {v}" for e, l, v in fields])
    msg = f"{title}\n\n{body}\n📅 时间: {now_shanghai()}"
    
    if TELEGRAM_BOT_TOKEN and TELEGRAM_CHAT_ID:
        try:
            requests.post(f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage", 
                          data={"chat_id": TELEGRAM_CHAT_ID, "text": msg, "parse_mode": "HTML"}, timeout=5)
        except: pass

# ================= 主流程 =================
def run_task():
    driver = None
    target_name = "未知" 
    server_id = "未知"
    login_ip = "Unknown"
    target_name = TARGET_NAME_CONFIG
    
    try:
        opts = Options()
        opts.add_argument("--headless=new")
        opts.add_argument("--no-sandbox")
        driver = webdriver.Chrome(options=opts, seleniumwire_options={'proxy': {'http': PROXY_URL, 'https': PROXY_URL}} if PROXY_URL else None)
        wait = WebDriverWait(driver, 25)

        # 0. 登入 IP 打印
        try:
            driver.get("https://api.ipify.org?format=json")
            login_ip = json.loads(driver.find_element(By.TAG_NAME, "body").text).get('ip', 'Unknown')
            print(f"🌐 登入 IP: {login_ip}")
        except: print("🌐 登入 IP: 无法获取")

        # 1. 登录
        print(f"🔑 正在登录: {EMAIL[:3]}***...")
        driver.get("https://greathost.es/login")
        wait.until(EC.presence_of_element_located((By.NAME,"email"))).send_keys(EMAIL)
        driver.find_element(By.NAME,"password").send_keys(PASSWORD)
        driver.find_element(By.CSS_SELECTOR,"button[type='submit']").click()
        wait.until(EC.url_contains("/dashboard"))

        # 2. 获取服务器列表并锁定
        res = fetch_api(driver, "/api/servers")
        server_list = res.get('servers', [])
        target_server = next((s for s in server_list if s.get('name') == target_name), None)
        
        if not target_server: raise Exception(f"未找到服务器 {target_name}")
        server_id = target_server.get('id')
        print(f"✅ 已锁定目标服务器: {target_name} (ID: {server_id})")
        
        # 3. 获取实时状态
        info = fetch_api(driver, f"/api/servers/{server_id}/information")
        real_status = info.get('status', 'unknown').lower()
        icon, status_name = STATUS_MAP.get(real_status, ["❓", real_status])
        status_disp = f"{icon} {status_name}"
        print(f"📋 状态核对: {target_name} | {status_disp}")

        # 4. 时间检查与冷却判定
        driver.get(f"https://greathost.es/contracts/{server_id}")
        time.sleep(2)
        
        # 【修正点】验证过的 renewal 接口，并增加 .get('contract') 层级
        contract_data = fetch_api(driver, f"/api/renewal/contracts/{server_id}")

        print(f"DEBUG: 原始合同数据 -> {str(contract_data)[:100]}...")
        # 兼容性处理：优先找 contract 里的数据，找不到则看顶层
        renewal_info = contract_data.get('contract', {}).get('renewalInfo') or contract_data.get('renewalInfo', {})
        before_h = calculate_hours(renewal_info.get('nextRenewalDate'))
        
        btn = wait.until(EC.presence_of_element_located((By.ID, "renew-free-server-btn")))
        btn_text = btn.text.strip()
        print(f"🔘 按钮状态: '{btn_text}' | 剩余: {before_h}h")
        
        if "Wait" in btn_text:
            m = re.search(r"Wait\s+(\d+\s+\w+)", btn_text)
            wait_time = m.group(1) if m else btn_text
            send_notice("cooldown", [
                ("📛", "服务器名称", target_name),
                ("🆔", "ID", f"<code>{server_id}</code>"),
                ("⏳", "冷却时间", wait_time),
                ("📊", "当前累计", f"{before_h}h"),
                ("🚀", "服务器状态", status_disp)
            ])
            return

        # 5. 执行续期
        print(f"🚀 正在执行续期 POST...")
        renew_res = fetch_api(driver, f"/api/renewal/contracts/{server_id}/renew-free", method="POST")
        
        is_success = renew_res.get('success', False)
        # 续期成功后，新日期确实是在 details 字段下
        after_date = renew_res.get('details', {}).get('nextRenewalDate')
        res_message = renew_res.get('message', '无返回消息')
         # 后台打印完整的响应信息，方便调试
        print(f"📡 续期响应结果: Success={is_success} | Message='{res_message}'")
        
        # 确保 after_h 在失败时不会变成 0
        if is_success and after_date:
            after_h = calculate_hours(after_date)
        else:
            after_h = before_h

        # 6. 判定并发送通知
        if is_success and after_h > before_h:
            send_notice("renew_success", [
                ("📛", "服务器名称", target_name),
                ("🆔", "ID", f"<code>{server_id}</code>"),
                ("⏰", "增加时间", f"{before_h} ➔ {after_h}h"),
                ("🚀", "服务器状态", status_disp),
                ("🌐", "登入 IP", f"<code>{login_ip}</code>")
            ])
        elif "5 d" in str(renew_res.get('message', '')) or (before_h > 108):
            send_notice("maxed_out", [
                ("📛", "服务器名称", target_name),
                ("🆔", "ID", f"<code>{server_id}</code>"),
                ("⏰", "剩余时间", f"{after_h}h"),
                ("🚀", "服务器状态", status_disp),
                ("💡", "提示", res_message),
                ("🌐", "登入 IP", f"<code>{login_ip}</code>")
            ])
        else:
            send_notice("renew_failed", [
                ("📛", "服务器名称", target_name),
                ("🆔", "ID", f"<code>{server_id}</code>"),
                ("⏰", "剩余时间", f"{before_h}h"),
                ("💡", "提示", res_message)
            ])

    except Exception as e:
        print(f"🚨 运行异常: {e}")
        send_notice("error", [("📛", "服务器名称", target_name), ("❌", "故障", f"<code>{str(e)[:100]}</code>")])
    finally:
        if driver: driver.quit()

if __name__ == "__main__":
    run_task()
